basic zoo structure for predict and train for Nobrainer models

Hoda1394 opened the pull request:
the brainy script is tested and working. ams needs some modifications. will add a specific one for kwyk.

satra commented on ams_inference.py:
```diff
@@ -0,0 +1,48 @@
+batch_size = 1
+
+# Load pre-trained model
+model_path = "<path_to_trained_model_folder>/trained-models/neuronets/ams/0.1.0/meningioma_T1wc_128iso_v1.h5"
```
shouldn't this be automated with trained-models?

Hoda1394 replied:
Here, it is supposed that the user already got the model from trained_models repo. so, I am not sure what you mean by automation here. should I add something in the zoo that gets the model?

satra replied:
as in the user should not have to do this.  knowing this is the inference model and version, the script should take care of this. 

the user in the ideal world would only be doing something like:

```
nobrainer-zoo predict [PREDICT_PARAMETERS if any] --input <file/bids_dataset> --model <org>.<model_name>.<version>  
nobrainer-zoo predict --help
```

if the user is a developer it should be something quite similar:

```
from nobrainerzoo.<org> import model_name

model = model_name(version=<version>)
# internally this should do all the necessary pieces for setting up the environment if necessary and running the model.
out = model.predict(input, **parameters)  
```

the exact details can be worked on, but this should be the pattern for every model in the zoo.

Hoda1394 replied:
I see, I will work toward this.

Hoda1394 commented:
@djarecka- here is the basic structure for models in nobrainer. currently, there is an issue with getting the data with datalad using the container. 
right now it creates the models repo in the current working directory.

Hoda1394 commented:
There are still some ToDo for braingen models.
kwyk prediction function should also be added to Nobrainer cli.

satra commented:
i would still suggest following an organization as suggested in the email for synthmorph, where each model has it's own org/directory/version

djarecka commented:
@Hoda1394 - just realized that I wrote comments before weekend, but didn't post it... I don't see any new commits, so hopefully they are still  relevant...

I have several question about this PR:
- I was not able to run an example, since it complains about `nobrainer-zoo_test.sif`. How do I get the image? Is it related to the `nobrainer_Docker_file`?
- I understand that each organization/model or version could have a different image, so we could put this to the `organization/model /version` directory
- there  is any reason why you run `download.py` in the singularity image? Do you want to download the model every single time you run every single time you call `gpu_run`
- there is anything that works only for `gpu` in the `gpu_run`?

Hoda1394 commented:
@djarecka - Some bugs remained that I fixed today, and I will push the changes and then we can discuss the questions based on the new changes. They are still valid.

Hoda1394 commented:
@djarecka - This is a base structure and works for brainy and ams models. some changes needed for kwyk and braingen. also for other models like synthmorph.
- you can pull the image from here [https://hub.docker.com/repository/docker/hodadock/nobrainer-zoo](https://hub.docker.com/repository/docker/hodadock/nobrainer-zoo) and put it in `zoo/nobrainerzoo/env/`. Actually, the cli should pull the proper image based on the model. it is still a TODO in my code.
- I changed the overall structure to `organization/model `. For `neuronet` models they are stored in `trained-models` repo that is cloned and related model is downloaded if not already.
- `download.py` downloads a model if it is not already downloaded. The reason that I am running it inside an image is that it needs `datalad` and `datalad` is provided by the image. maybe more efficient way is running the singularity image only if the model needs to be downloaded which means separating the checking part from download part.
- I implemented it only with singularity. docker commands need to be added for cpu. with changing the function names in the cli maybe it can be a flag `cpu/gpu`.

We can chat more when you are available.

djarecka commented:
> * you can pull the image from here https://hub.docker.com/repository/docker/hodadock/nobrainer-zoo and put it in `zoo/nobrainerzoo/env/`. Actually, the cli should pull the proper image based on the model. it is still a TODO in my code.
That's a good idea to pull a correct image for specific model, but I think for the future it might be better to request a name of the image to be part of the "model specification" and not be specified by you in the main function. This would give you more flexibility if other people want to add another models.

> * `download.py` downloads a model if it is not already downloaded. The reason that I am running it inside an image is that it needs `datalad` and `datalad` is provided by the image. maybe more efficient way is running the singularity image only if the model needs to be downloaded which means separating the checking part from download part.

IMO, `datalad` should be a requirement for the `zoo` repository. The image needs to have software to "train" or "predict" only.

> * I implemented it only with singularity. docker commands need to be added for cpu. with changing the function names in the cli maybe it can be a flag `cpu/gpu`.
 
So you will have a different command for `cpu`?

> We can chat more when you are available.
Sure, happy to chat!

Hoda1394 commented:
> That's a good idea to pull a correct image for specific model, but I think for the future it might be better to request a name of the image to be part of the "model specification" and not be specified by you in the main function. This would give you more flexibility if other people want to add another models.

The goal is not to ask it from the user. so each model that is integrated into the zoo should be associated with an image that provides a proper environment for training and prediction. but under the hood it can be set as a model spec. if it makes things easier.

> IMO, datalad should be a requirement for the zoo repository. The image needs to have software to "train" or "predict" only.

Not all models will be integrated and accessed through datalad . So we need it only for models that can be accessed through datalad. for example, models stored in `trained_models` repo.

> So you will have a different command for cpu?

Well, the answer is no, the predict and train command won't change. Here the main difference would be using singularity or docker depending on where to run the commands (HPC or local machine). So cpu/gpu keyword may not be a good choice. Do you have any suggestions for the implementation of this part?

djarecka commented:
> > That's a good idea to pull a correct image for specific model, but I think for the future it might be better to request a name of the image to be part of the "model specification" and not be specified by you in the main function. This would give you more flexibility if other people want to add another models.
> 
> The goal is not to ask it from the user. so each model that is integrated into the zoo should be associated with an image that provides a proper environment for training and prediction. but under the hood it can be set as a model spec. if it makes things easier.
I ment "user-developer", who is adding more models

> 
> > IMO, datalad should be a requirement for the zoo repository. The image needs to have software to "train" or "predict" only.
> 
> Not all models will be integrated and accessed through datalad . So we need it only for models that can be accessed through datalad. for example, models stored in `trained_models` repo.

So we can have datalad as an optional dependency, but it still looks to me that it would be better to keep it outside the main image

> > So you will have a different command for cpu?
> 
> Well, the answer is no, the predict and train command won't change. Here the main difference would be using singularity or docker depending on where to run the commands (HPC or local machine). So cpu/gpu keyword may not be a good choice. Do you have any suggestions for the implementation of this part?

in this case, we can have `--container_type` (or something) flag that takes `docker` or `singularity`.

Hoda1394 commented:
@djarecka - I fixed a bug and now it works.

Hoda1394 commented:
@satra this PR is ready to be merged.
